File: mobile/lib/core/auth/auth_session.dart
/// Người dùng trả về kèm token. Ánh xạ đúng `AuthDtos.AuthUserResponse` của backend Java.
class AuthUser {
  const AuthUser({
    required this.userId,
    required this.fullName,
    required this.email,
    required this.role,
  });

  final String userId;
  final String fullName;
  final String email;
  final String role;

  factory AuthUser.fromJson(Map<String, dynamic> json) => AuthUser(
        userId: json['userId'] as String,
        fullName: json['fullName'] as String,
        email: json['email'] as String,
        role: json['role'] as String,
      );

  Map<String, dynamic> toJson() => {
        'userId': userId,
        'fullName': fullName,
        'email': email,
        'role': role,
      };
}

/// Phiên đăng nhập đã lưu trên máy: token, hạn dùng, và người dùng.
class AuthSession {
  const AuthSession({
    required this.accessToken,
    required this.expiresAt,
    required this.user,
  });

  final String accessToken;

  /// Luôn giữ ở **UTC**. Backend trả `Instant` (ISO-8601, hậu tố `Z`); nếu để giờ máy thì một
  /// thiết bị đặt sai múi giờ sẽ tự cho là token còn hạn hoặc đã hết hạn sớm vài tiếng.
  final DateTime expiresAt;

  final AuthUser user;

  /// Khoảng lùi trước hạn.
  ///
  /// Token còn đúng 20 giây không dùng được: request bay đi, mạng 3G trong quán mất 2–3 giây,
  /// tới nơi thì token đã chết và người dùng nhận 401 giữa lúc đang đặt món. Coi như hết hạn
  /// sớm hơn một phút để phần gọi mạng luôn có token thật sự còn sống.
  static const Duration bienAnToan = Duration(minutes: 1);

  bool conHieuLuc(DateTime now) {
    return expiresAt.toUtc().subtract(bienAnToan).isAfter(now.toUtc());
  }

  /// KHÔNG in token ra.
  ///
  /// `toString()` của một object bị gọi ở những chỗ không ai ngờ: `print(session)` lúc gỡ lỗi,
  /// log của Flutter khi widget ném lỗi, và báo cáo sự cố gửi lên dịch vụ ngoài. Mặc định của
  /// Dart in tên lớp nên vốn đã an toàn, nhưng nó cũng vô dụng khi gỡ lỗi — nên viết lại thành
  /// bản có ích mà vẫn không lộ token, thay vì để người sau thêm token vào cho "dễ debug".
  @override
  String toString() =>
      'AuthSession(user: ${user.email}, role: ${user.role}, expiresAt: ${expiresAt.toIso8601String()})';

  factory AuthSession.fromJson(Map<String, dynamic> json) => AuthSession(
        accessToken: json['accessToken'] as String,
        expiresAt: DateTime.parse(json['expiresAt'] as String).toUtc(),
        user: AuthUser.fromJson(json['user'] as Map<String, dynamic>),
      );

  Map<String, dynamic> toJson() => {
        'accessToken': accessToken,
        'expiresAt': expiresAt.toUtc().toIso8601String(),
        'user': user.toJson(),
      };
}

File: mobile-rn/src/core/auth/authSession.ts
/** Người dùng trả về kèm token. Ánh xạ đúng `AuthDtos.AuthUserResponse` của backend Java. */
export interface AuthUser {
  readonly userId: string;
  readonly fullName: string;
  readonly email: string;
  readonly role: string;
}

/** Phiên đăng nhập đã lưu trên máy: token, hạn dùng, và người dùng. */
export interface AuthSession {
  readonly accessToken: string;
  /**
   * Luôn giữ ở **UTC**, dạng ISO-8601. Backend trả `Instant` với hậu tố `Z`; nếu để giờ máy thì
   * một thiết bị đặt sai múi giờ sẽ tự cho là token còn hạn hoặc đã hết hạn sớm vài tiếng.
   */
  readonly expiresAt: string;
  readonly user: AuthUser;
}

/**
 * Khoảng lùi trước hạn, tính bằng mili giây.
 *
 * Token còn đúng 20 giây không dùng được: request bay đi, mạng 3G trong quán mất 2–3 giây, tới
 * nơi thì token đã chết và người dùng nhận 401 giữa lúc đang đặt món. Coi như hết hạn sớm hơn
 * một phút để phần gọi mạng luôn có token thật sự còn sống.
 */
export const BIEN_AN_TOAN_MS = 60_000;

export function conHieuLuc(session: AuthSession, bayGio: Date): boolean {
  const han = Date.parse(session.expiresAt);
  if (Number.isNaN(han)) return false;
  return han - BIEN_AN_TOAN_MS > bayGio.getTime();
}

export function authUserTuJson(json: unknown): AuthUser {
  const o = json as Record<string, unknown>;
  return {
    userId: o.userId as string,
    fullName: o.fullName as string,
    email: o.email as string,
    role: o.role as string,
  };
}

export function authSessionTuJson(json: unknown): AuthSession {
  const o = json as Record<string, unknown>;
  const han = new Date(o.expiresAt as string);
  if (Number.isNaN(han.getTime())) throw new Error('expiresAt không đọc được');
  return {
    accessToken: o.accessToken as string,
    expiresAt: han.toISOString(),
    user: authUserTuJson(o.user),
  };
}

/**
 * KHÔNG in token ra.
 *
 * Object bị đưa vào chuỗi ở những chỗ không ai ngờ: `console.log(session)` lúc gỡ lỗi, log của
 * React Native khi component ném lỗi, và báo cáo sự cố gửi lên dịch vụ ngoài. Khác với Dart, mặc
 * định của JavaScript là in HẾT mọi trường — tức token lộ ngay từ lần `console.log` đầu tiên.
 * Nên ở đây phải có một hàm mô tả có ích để không ai phải log nguyên object.
 */
export function moTaSession(session: AuthSession): string {
  return `AuthSession(user: ${session.user.email}, role: ${session.user.role}, expiresAt: ${session.expiresAt})`;
}
